src/nupic/algorithms/AnomalyLikelihood.cpp

```diff
@@ -0,0 +1,592 @@
+  :param averagingWindow: integer number of records to average over
+  :param skipRecords: integer specifying number of records to skip when
+                      estimating distributions. If skip records are >=
+                      len(anomalyScores), a very broad distribution is returned
+                      that makes everything pretty likely.
+  :param verbosity: integer controlling extent of printouts for debugging
+
+                      0 = none
+                      1 = occasional information
+                      2 = print every record
+
+  :returns: 3-tuple consisting of:
+
+            - likelihoods
+
+              numpy array of likelihoods, one for each aggregated point
+
+            - avgRecordList
+
+              list of averaged input records
+
+            - params
+              a small JSON dict that contains the state of the estimator
+
+  **/
+  if (verbosity > 1) {
+    cout << "In estimateAnomalyLikelihoods."<<endl;
+    cout << "Number of anomaly scores:" <<  anomalyScores.size() << endl;
+    cout << "Skip records="<<  skipRecords << endl;
+//    print("First 20:", anomalyScores[0:min(20, len(anomalyScores))])
+  }
+
+  assert(anomalyScores.size() >= 0); // "Must have at least one anomalyScore"
+  auto dataValues = anomalyScores; //FIXME the "data" should be anomaly scores, or raw values? 
+
+  // Estimate the distribution of anomaly scores based on aggregated records
+  if (dataValues.size()  <= skipRecords) {
+    this->distribution = nullDistribution();
+  } else {
+    dataValues.erase(dataValues.begin(), dataValues.begin() + skipRecords);// remove first skipRecords
+    this->distribution = estimateNormal(dataValues);
+  
+    /* HACK ALERT! The CLA model currently does not handle constant metric values //TODO can this be removed now? 4/2017
+     very well (time of day encoder changes sometimes lead to unstable SDR's
+     even though the metric is constant). Until this is resolved, we explicitly
+     detect and handle completely flat metric values by reporting them as not
+     anomalous.
+    */
+      auto metricDistribution = estimateNormal(dataValues, false);
+      if (metricDistribution.variance < 1.5e-5) {
+        this->distribution = nullDistribution();
+      } //end of hack
+  }
+
+  // Estimate likelihoods based on this distribution
+  vector<Real> likelihoods;
+  for (auto s : dataValues) {
+    likelihoods.push_back(tailProbability(s));
+  }
+  // Filter likelihood values
+  auto filteredLikelihoods = filterLikelihoods_(likelihoods);
+
+
+/*  if verbosity > 1:
+    print("Discovered params=")
+    print(params)
+    print("Number of likelihoods:", len(likelihoods))
+    print("First 20 likelihoods:", (
+      filteredLikelihoods[0:min(20, len(filteredLikelihoods))] ))
+    print("leaving estimateAnomalyLikelihoods")
+*/
+
+  return filteredLikelihoods;
+}
+
+vector<Real> circularBufferToVector(boost::circular_buffer<Real> cb) const {
+  cb.linearize();
+  auto d1 = cb.array_one();
```

Comment:
I need to get data as a vector from the sliding window, for the `boost::circular_buffer` I need this hack. This is what I was mentioning in #1276